### PR TITLE
Add Student Forums Usage Report to Instructor Dashboard

### PR DESCRIPTION
**Pull request by azimpradhan**
Many instructors have asked to be able to include
forum participation as part of students' grades.
Some instructors may be interested in using this
report for other reasons, and some may be
interested in using it for grading. At the
moment there's no way for instructors to get
data on forum usage per student.

This commit allows instructor to request a report
showing forum usage per student. The
report displays the student's username, the total
number of posts the student has made in the course,
and the number of points their posts have earned.

Tests have been created for this commit and can be
run as follows:
python ./manage.py lms test --verbosity=1
lms/djangoapps/instructor_task/tests/test_tasks_helper.py:TestInstructorStudentForumsReport --traceback --settings=test
python ./manage.py lms test --verbosity=1 lms/djangoapps/instructor_task/tests/test_tasks.py:TestStudentForumsUsageInstructorTask --traceback --settings=test
python ./manage.py lms test --verbosity=1 lms/djangoapps/instructor_task/tests/test_api.py:InstructorTaskCourseSubmitTest.test_submit_student_forums_usage_task --traceback --settings=test
python ./manage.py lms test --verbosity=1 lms/djangoapps/instructor/tests/test_api.py:TestInstructorAPILevelsDataDump.test_get_student_forums_usage_success --traceback --settings=test
python ./manage.py lms test --verbosity=1 lms/djangoapps/instructor/tests/test_api.py:TestInstructorAPILevelsDataDump.test_get_student_forums_usage_already_running --traceback --settings=test

### CODE REVIEW

**Comment by azimpradhan** [comment]
tagging @caijim and @kluo

File: lms/djangoapps/instructor/utils.py [hunk 1]
```diff
@@ -155,3 +161,52 @@ def ora2_data_queries():
...
+            result['points'],
+        ] for result in results
+    ]
+    header = ['username', 'posts', 'points']
+    return header, parsed_results
+
+
+def generate_student_forums_query(course_id):
+    """
+    generates aan aggregate query for student data which can be executed using pymongo
```
Comment (caijim) [inline lms/djangoapps/instructor/utils.py]: an

File: lms/djangoapps/instructor_task/tasks.py [hunk 2]
```diff
@@ -176,3 +177,13 @@ def get_ora2_responses(entry_id, xmodule_instance_args):
     action_name = ugettext_noop('generated')
     task_fn = partial(push_ora2_responses_to_s3, xmodule_instance_args)
     return run_main_task(entry_id, task_fn, action_name)
+
+
+@task(base=BaseInstructorTask, routing_key=settings.ORA2_RESPONSES_DOWNLOAD_ROUTING_KEY)
```
Comment (caijim) [inline lms/djangoapps/instructor_task/tasks.py]: Same as what Kevin said on my PR, add a new key in

lms/envs/aws.py
lms/envs/common.py

and set it to HIGH_MEM_QUEUE

**Comment by caijim** [comment]
looks good to me otherwise

File: lms/djangoapps/instructor/management/commands/collect_student_forums_data.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+"""
+Command to retrieve aggregated course forums data in a .csv
```
Comment (kluo) [inline lms/djangoapps/instructor/management/commands/collect_student_forums_data.py]: you mean student?

File: lms/djangoapps/instructor/management/commands/collect_student_forums_data.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+from instructor.utils import collect_student_forums_data
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.locations import SlashSeparatedCourseKey
+
+from django.core.management.base import BaseCommand, CommandError
+
+
+class Command(BaseCommand):
+    """
+    Query aggregated open assessment data, write to .csv
```
Comment (kluo) [inline lms/djangoapps/instructor/management/commands/collect_student_forums_data.py]: student forums

File: lms/djangoapps/instructor/management/commands/collect_ora2_data.py [hunk 5]
```diff
@@ -1,3 +1,6 @@
+"""
```
Comment (kluo) [inline lms/djangoapps/instructor/management/commands/collect_ora2_data.py]: this files changes are not related to the PR

File: lms/djangoapps/instructor/management/commands/collect_student_forums_data.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                             action='store', dest='output_dir', default=None,
+                             help="Write output to a directory rather than stdout"),
+    )
+
+    def handle(self, *args, **options):
+        if not args:
+            raise CommandError("Course ID must be specified to fetch data")
+
+        try:
+            course_id = SlashSeparatedCourseKey.from_deprecated_string(args[0].decode('utf-8'))
```
Comment (kluo) [inline lms/djangoapps/instructor/management/commands/collect_student_forums_data.py]: the `decode('utf-8')` is not needed

File: lms/djangoapps/instructor/utils.py [hunk 7]
```diff
@@ -155,3 +161,52 @@ def ora2_data_queries():
...
+        raise
+
+    parsed_results = [
+        [
+            result['_id'],
+            result['posts'],
+            result['points'],
+        ] for result in results
+    ]
+    header = ['username', 'posts', 'points']
```
Comment (kluo) [inline lms/djangoapps/instructor/utils.py]: Maybe uppercase will look nicer?

File: lms/djangoapps/instructor/views/api.py [hunk 8]
```diff
@@ -2067,6 +2067,31 @@ def extract_user_info(user):
 @ensure_csrf_cookie
 @cache_control(no_cache=True, no_store=True, must_revalidate=True)
 @require_level('staff')
+def get_student_forums_usage(request, course_id):
+    """
+    Pushes a Celery task which will aggregate forum usage for a course into a .csv
```
Comment (kluo) [inline lms/djangoapps/instructor/views/api.py]: student forum usage

File: lms/djangoapps/instructor/views/api_urls.py [hunk 9]
```diff
@@ -122,7 +122,12 @@
     url(r'^gradebook$',
         'instructor.views.api.spoc_gradebook', name='spoc_gradebook'),
 
+    # Collect student forum data
```
Comment (kluo) [inline lms/djangoapps/instructor/views/api_urls.py]: forum[s]

File: lms/djangoapps/instructor_task/api.py [hunk 10]
```diff
@@ -238,7 +239,6 @@ def submit_get_student_responses(request, course_key):
 
     return submit_task(request, task_type, task_class, course_key, task_input, task_key)
 
-
```
Comment (kluo) [inline lms/djangoapps/instructor_task/api.py]: why is this being removed?

File: lms/djangoapps/instructor_task/api.py [hunk 11]
```diff
@@ -250,7 +250,6 @@ def submit_ora2_request_task(request, course_key):
...
```
Comment (kluo) [inline lms/djangoapps/instructor_task/api.py]: why is this being removed?

File: lms/djangoapps/instructor_task/api.py [hunk 12]
```diff
@@ -263,3 +262,16 @@ def submit_calculate_students_features_csv(request, course_key, features):
     task_key = ""
 
     return submit_task(request, task_type, task_class, course_key, task_input, task_key)
+
+
+# pylint: disable-msg=C0103
```
Comment (kluo) [inline lms/djangoapps/instructor_task/api.py]: what is this for?

File: lms/djangoapps/instructor_task/tasks_helper.py [hunk 13]
```diff
@@ -648,12 +648,22 @@ def push_student_responses_to_s3(_xmodule_instance_args, _entry_id, course_id, _
 
 
 def push_ora2_responses_to_s3(_xmodule_instance_args, _entry_id, course_id, _task_input, action_name):
+    return push_csv_responses_to_s3(collect_ora2_data, u'ORA2_responses', course_id, action_name)
```
Comment (kluo) [inline lms/djangoapps/instructor_task/tasks_helper.py]: this function still needs its original docstring

File: lms/djangoapps/instructor_task/tasks_helper.py [hunk 14]
```diff
@@ -648,12 +648,22 @@ def push_student_responses_to_s3(_xmodule_instance_args, _entry_id, course_id, _
...
-    Collect ora2 responses and upload them to S3 as a CSV
+    Generate student forums report and upload it to s3 as a CSV
     """
+    return push_csv_responses_to_s3(collect_student_forums_data, u'student_forums', course_id, action_name)
 
-    start_time = datetime.now(UTC)
 
+def push_csv_responses_to_s3(csv_fn, filename, course_id, action_name):
+    """
+    Helper Method that generates a generates a report and uploads it as a csv
```
Comment (kluo) [inline lms/djangoapps/instructor_task/tasks_helper.py]: Need to clean this up

File: lms/djangoapps/instructor_task/tasks_helper.py [hunk 15]
```diff
@@ -695,11 +705,11 @@ def update_task_progress():
...
 
-    report_store = ReportStore.from_config()
-    report_store.store_rows(
+    upload_csv_to_report_store(
+        rows,
+        filename,
         course_id,
-        u'{}_ORA2_responses_{}.csv'.format(course_id_string, timestamp_str),
-        rows
+        datetime.now(UTC),
```
Comment (kluo) [inline lms/djangoapps/instructor_task/tasks_helper.py]: Use start_time on line 666

File: lms/djangoapps/instructor_task/tests/test_tasks.py [hunk 16]
```diff
@@ -476,3 +478,30 @@ def test_ora2_runs_task(self):
             task_fn = partial(push_ora2_responses_to_s3, task_xmodule_args)
 
             mock_main_task.assert_called_once_with_args(task_entry.id, task_fn, action_name)
+
+
+class TestStudentForumsUsageInstructorTask(TestInstructorTasks):
+    """Tests instructor tasks that fetches student forums data."""
```
Comment (kluo) [inline lms/djangoapps/instructor_task/tests/test_tasks.py]: task with no s

File: lms/static/coffee/src/instructor_dashboard/data_download.coffee [hunk 17]
```diff
@@ -121,6 +121,8 @@ class DataDownload
             @$reports_request_response_error.text gettext("Error getting student responses. Please try again.")
           else if e.target.name == 'ora2-response-btn'
             @$reports_request_response_error.text gettext("Error getting ORA2 responses. Please try again.")
+          else if e.target.name == 'student-forums-btn'
+            @$reports_request_response_error.text gettext("Error getting Student Forum data. Please try again.")
```
Comment (kluo) [inline lms/static/coffee/src/instructor_dashboard/data_download.coffee]: forum[s]

File: lms/templates/instructor/instructor_dashboard_2/data_download.html [hunk 18]
```diff
@@ -44,6 +44,7 @@
       <input type="button" name="calculate-grades-csv" value="${_("Generate Grade Report")}" data-endpoint="${ section_data['calculate_grades_csv_url'] }"/>
       <input type="button" name="get-student-responses" value="${_("Get Student Responses Report")}" data-endpoint="${ section_data['get_student_responses_url'] }"/>
       <input type="button" name="ora2-response-btn" value="${_("Download Ora2 Responses")}" data-endpoint="${ section_data['get_ora2_responses_url'] }"/>
+      <input type="button" name="student-forums-btn" value="${_("Generate Student Forums Usage")}" data-endpoint="${ section_data['get_student_forums_usage_url']}"/>
```
Comment (kluo) [inline lms/templates/instructor/instructor_dashboard_2/data_download.html]: I think Student Forums Usage Report is clearer

**Comment by kluo** [comment]
I think Jim's course forums PR will probably get merged first, and there is lot of changes to common files between that PR and this one, so once Jim's PR merges you might have to resolve some conflicts on this PR when you rebase on master

**Comment by caijim** [comment]
Went through it and did a spot check and things look good to me

**Comment by azimpradhan** [comment]
can I get a +1 before I merge, @kluo

File: lms/djangoapps/instructor_task/api.py [hunk 19]
```diff
@@ -276,3 +277,15 @@ def submit_calculate_students_features_csv(request, course_key, features):
     task_key = ""
 
     return submit_task(request, task_type, task_class, course_key, task_input, task_key)
+
+
+def submit_stud_forums_usage_task(request, course_key):
```
Comment (kluo) [inline lms/djangoapps/instructor_task/api.py]: did you mean to abbreviate student here? seems inconsistent with everywhere else

Comment (azimpradhan) [inline lms/djangoapps/instructor_task/api.py]: Yes, the reason I did that is because I get a pylint error for the method name being too long. It is this particular method name change is consistent throughout the code.

**Comment by kluo** [comment]
Just one comment, otherwise LGTM

File: lms/djangoapps/instructor_task/api.py (hunk 19 above)
Comment (kluo) [inline lms/djangoapps/instructor_task/api.py]: Hm, the other task functions in this file are longer than this. @stvstnfrd is the pylint error worth truncating student here?

File: lms/djangoapps/instructor/management/commands/collect_student_forums_data.py [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+class Command(BaseCommand):
+    """
+    Retrieve aggregated student forums data, write to .csv
+    """
+
+    help = ("Usage: collect_course_forums_data <course_id> --output-dir=<output_dir>")
+    args = "<course_id>"
+    option_list = BaseCommand.option_list + (
+        optparse.make_option('-o', '--output-dir',
```
Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/collect_student_forums_data.py]: This block deviates from the community style guide [1].

[1] https://github.com/edx/edx-platform/wiki/Python-Guidelines

Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/collect_student_forums_data.py]: Sorry, "in relation to splitting long lines".

File: lms/envs/aws.py [hunk 21]
```diff
@@ -437,6 +437,9 @@
 # Course Forums Download
 COURSE_FORUMS_DOWNLOAD_ROUTING_KEY = HIGH_MEM_QUEUE
 
+#Student Forums Download
+STUDENT_FORUMS_DOWNLOAD_ROUTING_KEY = HIGH_MEM_QUEUE
```
Comment (stvstnfrd) [inline lms/envs/aws.py]: This should actually try to load the value from `ENV`, falling back to the value from `lms/envs/common.py`.

Comment (stvstnfrd) [inline lms/envs/aws.py]:
```python
STUDENT_FORUMS_DOWNLOAD_ROUTING_KEY = ENV_TOKENS.get(
    'STUDENT_FORUMS_DOWNLOAD_ROUTING_KEY',
    STUDENT_FORUMS_DOWNLOAD_ROUTING_KEY,
)
```

File: lms/djangoapps/instructor/management/commands/collect_course_forums_data.py [hunk 22]
```diff
@@ -33,7 +33,7 @@ def handle(self, *args, **options):
         except InvalidKeyError:
             raise CommandError("The course ID given was invalid")
 
-        file_name = "{course_id}-course-forums.csv".format(course_id=course_id.replace("/", "-"))
+        file_name = "{course_id}-course-forums.csv".format(course_id=str(course_id).replace("/", "-"))
```
Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/collect_course_forums_data.py]: Creating a `CourseKey` from a deprecated string and then reconverting it back to a string via `str` (and not `to_deprecated_string` or even `unicode`) doesn't feel right.

If we need the string value, can't we just use the value from `args[0]`?

`str(SlashSeparatedCourseKey(args[0]))` _may_ be equivalent to `args[0]` for now, but I don't think we can always guarantee this to be true.

File: lms/djangoapps/instructor/management/commands/collect_student_forums_data.py [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+Command to retrieve aggregated student forums data in a .csv
+"""
+import csv
+import optparse
+import os
+
+from instructor.utils import collect_student_forums_data
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.locations import SlashSeparatedCourseKey
+from django.core.management.base import BaseCommand, CommandError
```
Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/collect_student_forums_data.py]: Can you group these imports according to the style guide [1]?

[1] https://github.com/edx/edx-platform/wiki/Python-Guidelines#imports-order

File: lms/djangoapps/instructor/management/commands/collect_student_forums_data.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class Command(BaseCommand):
+    """
+    Retrieve aggregated student forums data, write to .csv
+    """
+
+    help = ("Usage: collect_course_forums_data <course_id> --output-dir=<output_dir>")
+    args = "<course_id>"
```
Comment (stvstnfrd) [inline lms/djangoapps/instructor/management/commands/collect_student_forums_data.py]: Nit: Single quotes for non-interpolated strings here and throughout.

File: lms/djangoapps/instructor/views/instructor_dashboard.py [hunk 25]
```diff
@@ -328,6 +328,7 @@ def _section_data_download(course, access):
         'list_report_downloads_url': reverse('list_report_downloads', kwargs={'course_id': course_key.to_deprecated_string()}),
         'calculate_grades_csv_url': reverse('calculate_grades_csv', kwargs={'course_id': course_key.to_deprecated_string()}),
         'get_student_responses_url': reverse('get_student_responses', kwargs={'course_id': course_key.to_deprecated_string()}),
+        'get_student_forums_usage_url': reverse('get_student_forums_usage', kwargs={'course_id': course_key.to_deprecated_string()}),
```
Comment (stvstnfrd) [inline lms/djangoapps/instructor/views/instructor_dashboard.py]: New code should _almost_ *never* call `to_deprecated_string` or `from_deprecated_string`.

I know the other code here still follows this pattern, but that's because we're behind upstream.  They have already converted their legacy invocations [1]; we should do the same for our new ones.

[1] https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/views/instructor_dashboard.py#L161
